### ISSUE TITLE
Issues about Citation Key generation of Chinese articles

### ISSUE THREAD

**Issue by FDUZS** [retorquere/zotero-better-bibtex#2089]
I'd like to make the citation key shorter to be the same as the Google Scholar flavor.

```
@article{bush1945we,
  title={As we may think},
  author={Bush, Vannevar and others},
  journal={The atlantic monthly},
  volume={176},
  number={1},
  pages={101--108},
  year={1945}
}
```

After setting the Citation key format as `[auth:lower][year][veryshorttitle:lower]`, it works well on English articles but not the Chinese ones. The latter of which citation key is still too long.

I try to use `jieba` (with `jieba` filter in citekey patterns enabled) to do the job but still in vain. And the following document seems kind of outdated.

> jieba: word segmentation for Chinese items. Uses substantial memory; must be enabled under Preferences -> Better BibTeX -> Advanced -> Citekeys

There is an extra issue I want to report. [This article](http://www.lis.ac.cn/CN/10.13266/j.issn.0252-3116.2021.18.015) in my library of which the title is `信息茧房研究综述`, while the automatically generated citation key is `zhu2021xinxichongfangyanjiuzongshu` which is not correct. Let me explain it further. `信` corresponds to `xin`, `息` to `xi`, but there is no such `chong` pronunciation of `茧` (should be `jian` instead).

**Comment by duncdrum** [comment]
@FDUZS try the syntax from [this post](https://github.com/retorquere/zotero-better-bibtex/issues/1678#issuecomment-711795652) `[Auth][Title:select=1,1:nopunct:split-ideographs:select=1,3][year]` to get shorter titles. 

> but there is no such chong pronunciation …

this is incorrect. e.g. `虫/蟲` is transliterated as `chong` in PRC pinyin. 

> … of 茧 (should be jian instead)

That `jieba` picks the very rare pronunciation variant `chong` for `茧` instead of the more widely used `jian` is pretty much a bug there, you should open an issue with them.

**Comment by FDUZS** [comment]
@duncdrum thanks for answering.

> this is incorrect. e.g. 虫/蟲 is transliterated as chong in PRC pinyin.

I may make myself misunderstood. what I mean is `chong` does not correspond to the `茧` character.

---

I have made [some tests](https://gist.github.com/FDUZS/d86ca69d6a2cdd97384c854499f79c51) to see what role of these `filter` plays in citation key generation respectively. and my final (maybe) format is `[auth:lower][year][Title:split-ideographs:nopunctordash:skipwords:select=1,3:lower]`.

though it's still like a black box to me, some implications may be made:

- `Title` works just like a charm compared to `title`, and I can find any document about it, or am I just missing sth?
- `:lower` filter does not work with Chinese articles when using `Title`
- the order of `filter` matters (?)
- multiple `select` filters can work with English articles but only one with Chinese ones (?)

well, that's all for now for me.
btw, which `jieba` library does this extension use so I can file an issue with them.

**Comment by retorquere** [comment]
> `Title` works just like a charm compared to `title`, and I can't find any document about it, or am I just missing sth?

https://retorque.re/zotero-better-bibtex/citing/#direct-access-to-unprocessed-fields

> `:lower` filter does not work with Chinese articles when using `Title`

I don't know Chinese; I don't know what it means to lowercase such characters.

> the order of `filter` matters (?)

Correct, but I don't see how that could be different. You pass data from one filter to another. 
 
> multiple `select` filters can work with English articles but only one with Chinese ones (?)

most filters just don't know about Chinese characters. `lower` for example just uses the javascript `toLowerCase` function.
 
> well, that's all for now for me. btw, which `jieba` library does this extension use so I can file an issue with them.

https://github.com/ooooevan/jieba

**Comment by FDUZS** [comment]
@retorquere thanks! I did miss the `Title` thing.

> You pass data from one filter to another.

I get it. It's like the pipeline in bash.

> what it means to lowercase such characters.

The widely used Chinese input method is [Pinyin](https://en.wikipedia.org/wiki/Pinyin). Like the example above, the Pinyin (without tones) of `信 息 茧 房` is `Xin Xi Jian Fang` respectively. To lowercase Pinyin, I think there will be no differences with English words.

Speaking of this, with the citation key format remaining the same (`[auth:lower][year][Title:split-ideographs:nopunctordash:skipwords:select=1,3:lower]`) and the `Force citation key to plain text` setting unchecked, the full citation key is `zhu2021信息茧`. If checked (which is the default option), it will be converted to `zhu2021XinXiChong` automatically and it's partly wrong as I said. So, maybe it's related to this setting?

More interestingly, even the setting is unchecked, the former part of the key is `zhu` (which is the Pinyin of `朱`, a Chinese surname) instead of `朱`. After substituting `auth` with `Auth`, when unchecked, the former part of the key is `朱`; but then, even with that setting checked and `[Auth:lower]` set, it will still be uppercase `Zhu`. Could this relate to the difference between `auth` and `Auth`, `title` and `Title`? 

To summarize, with `Auth` or `Title`, the `lower` filter will fail; while with `title`, the `split-ideographs` will fail; and there may be some problems with the `Force citation key to plain text` functionality if as described above.

**Comment by retorquere** [comment]
> The widely used Chinese input method is [Pinyin](https://en.wikipedia.org/wiki/Pinyin). Like the example above, the Pinyin (without tones) of `信 息 茧 房` is `Xin Xi Jian Fang` respectively. To lowercase Pinyin, I think there will be no differences with English words.

Then `Title:fold=chinese:lower` might do what you need. For no other reason than that it was requested first, the default is `japanese` if the language field is not set on the item, but if the item has chinese as its language (or some forms like `zh` which are also recognized), `Title:fold` should do it.

> Speaking of this, with the citation key format remaining the same (`[auth:lower][year][Title:split-ideographs:nopunctordash:skipwords:select=1,3:lower]`) and the `Force citation key to plain text` setting unchecked, the full citation key is `zhu2021信息茧`. If checked (which is the default option), it will be converted to `zhu2021XinXiChong` automatically and it's partly wrong as I said. So, maybe it's related to this setting?

Which "this" do you mean? But by and large, I won't be able to say much useful about transliteration. I use libraries and trust people like @duncdrum who do know Chinese.

> More interestingly, even the setting is unchecked, the former part of the key is `zhu` (which is the Pinyin of `朱`, a Chinese surname) instead of `朱`. After substituting `auth` with `Auth`, when unchecked, the former part of the key is `朱`; but then, even with that setting checked and `[Auth:lower]` set, it will still be uppercase `Zhu`. Could this relate to the difference between `auth` and `Auth`, `title` and `Title`?

More likely related to `fold` defaulting to jp. The only difference between `auth` and `Auth` is that the former has the default transliteration applied and non-ascii removed.

**Comment by FDUZS** [comment]
My final citation key format is `[auth:lower][year][Title:fold=chinese:lower:nopunctordash:select=1,3]`, and the generated key will be like `bush1996aswemay` & `zhu2021xinxichong` which perfectly suits my need.

Thanks for your help and I'll close this issue.
